storage/allocator.go

```diff
@@ -273,15 +279,27 @@ func (a allocator) removeTarget(existing []proto.Replica) (proto.Replica, error)
 	return worst.repl, nil
 }
 
-// shouldRebalance returns whether the specified store is overweight
-// according to the cluster mean and should rebalance a range.
-func (a allocator) shouldRebalance(s *proto.StoreDescriptor) bool {
-	sl := a.storePool.getStoreList(*s.CombinedAttrs(), a.deterministic)
-
+// shouldRebalance returns whether the specified store should attempt to
+// rebalance a replica to another store.
+func (a allocator) shouldRebalance(storeID proto.StoreID) bool {
+	// In production, add some random jitter to shouldRebalance.
+	if !a.options.Deterministic && a.randGen.Float32() > rebalanceShouldRebalanceChance {
+		return false
+	}
+	storeDesc := a.storePool.getStoreDescriptor(storeID)
+	if storeDesc == nil {
+		if log.V(1) {
+			log.Warningf(
+				"shouldRebalance couldn't find store with id %d in StorePool",
+				storeID)
+		}
+		return false
+	}
+	sl := a.storePool.getStoreList(*storeDesc.CombinedAttrs(), a.options.Deterministic)
 	if sl.used.mean < minFractionUsedThreshold {
-		return s.Capacity.RangeCount > int32(math.Ceil(sl.count.mean))
+		return storeDesc.Capacity.RangeCount > int32(math.Ceil(sl.count.mean))
```

Comment:
Yeah, I think we should be using `rebalanceFromMean` for all of these comparisons to the mean. Right now we won't consider a node to be a rebalance destination unless it is below the mean by `rebalanceFromMean`; we shouldn't consider a node as a rebalance source unless it is also above the mean by `rebalanceFromMean`.